config-controller/internal/controller/policy_controller.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/protoadapt"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	storage "github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/roxctl/common"
+	"github.com/stackrox/rox/roxctl/common/auth"
+	roxctlIO "github.com/stackrox/rox/roxctl/common/io"
+	"github.com/stackrox/rox/roxctl/common/logger"
+	"github.com/stackrox/rox/roxctl/common/printer"
+
+	configstackroxiov1alpha1 "github.com/stackrox/rox/config-controller/api/v1alpha1"
+)
+
+// PolicyReconciler reconciles a Policy object
+type PolicyReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	conn   *grpc.ClientConn
+}
+
+//+kubebuilder:rbac:groups=config.stackrox.io,resources=policies,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=config.stackrox.io,resources=policies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=config.stackrox.io,resources=policies/finalizers,verbs=update
+
+func (r *PolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	rlog := log.FromContext(ctx)
+	rlog.Info("Reconciling", "namespace", req.Namespace, "name", req.Name)
+
+	// Get the policy CR
+	policyCR := &configstackroxiov1alpha1.Policy{}
+	if err := r.Client.Get(ctx, req.NamespacedName, policyCR); err != nil {
+		if k8serr.IsNotFound(err) {
+			// Must have been deleted
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("Failed to get policy: namespace=%s, name=%s", req.Namespace, req.Name)
+	}
+
+	desiredState := &storage.Policy{}
+	if err := protojson.Unmarshal([]byte(policyCR.Spec.Policy), protoadapt.MessageV2Of(desiredState)); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "Failed to parse policy JSON")
+	}
+
+	// GET policy from Central
+	svc := v1.NewPolicyServiceClient(r.conn)
+	allPolicies, err := svc.ListPolicies(ctx, &v1.RawQuery{})
+
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "Failed to list all policies")
+	}
+
+	for _, policy := range allPolicies.Policies {
+		if policy.Name == req.Name {
```

Comment:
This would fail if someone has renamed a policy  controlled in git on Central - which means we should disallow such updates. Correct? (we have an ongoing slack discussion on this cc: @dvail @mansursyed )

Reply:
If someone renames a policy via the UI when that policy is actually controlled via gitops, this is what I would imagine would happen upon reconciliation:

1. The policy would be recreated with the old name based on what's in the gitops repo.
2. The renamed policy would be deleted since it's marked as managed by gitops but does not exist in the gitops repo.

This is not an error imo.

Reply:
PostPolicy succeeds only if the ID field of the policy being created is empty in the request (aka in git in our case). 
 https://github.com/stackrox/stackrox/blob/0e267bc6db4983704b40b3e7dcc94d8f832c157b/central/policy/service/service_impl.go#L289

Reply:
The rule of thumb for REST APIs is a POST creates a new object, and a PUT takes a complete existing object (including its ID) and updates it. 
PUT also usually have the ID of the object to be updated as a URL parameter, which should match the ID in the payload.

It is possible for POST to take the ID of the object to be created, if the system allows users to choose there own ID, but I would expect a POST with an ID that collides with an existing ID to fail in that case.

Reply:
Sounds like there are two ways for the controller to handle policy CRs that have an ID set but don't have a corresponding policy in the DB:

1. Reject the policy from Central
2. Ignore the ID

I think I lean towards option 1.  wdyt?

Reply:
😈 's advocate here:

If a org uses the same repo for policies, and promotes them through test, stage, and prod environments, at what point do the policy IDs get checked into the repo ?

In other words, a version policy eventually could have an ID in stage, but then when they want to later add it to prod, our controller in prod would reject it because it has an ID but isn't in the database yet?

Alternate suggest, just keep IDs out of the CR versions forever, and only match on the exact names.

Reply:
> just keep IDs out of the CR versions forever

Reply:
Captured this as a requirement in https://issues.redhat.com/browse/ROX-25503